internal/app/machined/pkg/controllers/config/node_labels.go

```diff
@@ -0,0 +1,334 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package config
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/go-pointer"
+	"go.uber.org/zap"
+
+	"github.com/talos-systems/talos/pkg/kubernetes"
+	talosconfig "github.com/talos-systems/talos/pkg/machinery/config"
+	"github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1/machine"
+	"github.com/talos-systems/talos/pkg/machinery/resources/config"
+	"github.com/talos-systems/talos/pkg/machinery/resources/k8s"
+)
+
+// NodeLabelsK8sClientIface is the minimal interface needed fo manage the node labels.
+type NodeLabelsK8sClientIface interface {
+	Close() error
+	GetNodeLabels(ctx context.Context, nodeName string) (map[string]string, error)
+	UpdateNodeLabels(ctx context.Context, nodeName string, labelsToSet map[string]string, labelsToRemove map[string]struct{}) (err error)
+}
+
+// NodeLabelsController manages config.NodeLabels based on configuration.
+type NodeLabelsController struct {
+	GetK8sClientFunc func(ctx context.Context, r controller.Runtime, cfgProvider talosconfig.Provider) (NodeLabelsK8sClientIface, error)
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *NodeLabelsController) Name() string {
+	return "config.NodeLabelsController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *NodeLabelsController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
```

Comment:
such controller should never be hanging on the machine config itself, but rather on an interim resource: MachineConfig -> NodeLabelsConfig -> ...